test/mongoose_subdomain_core_SUITE.erl

```diff
@@ -0,0 +1,567 @@
+                 get_list_of_subdomain_collisions()),
+    no_domain_collisions().
+
+detects_domain_subdomain_collisions(_Config) ->
+    %% there are two possible domain/subdomain names collisions:
+    %%   1) domain/subdomain_pattern pair produces the same subdomain name as another
+    %%      existing domain
+    %%   2) FQDN subdomain is the same as some registered domain
+    %%
+    %% the naive domain/subdomain registration rejection is probably a bad option:
+    %%   * domains and subdomains ETS tables are managed asynchronously, in addition to
+    %%     that subdomains patterns registration is done async as well. this all leaves
+    %%     room for a various race conditions if try to just make a verification and
+    %%     prohibit domain/subdomain registration in case of any collisions.
+    %%   * the only way to avoid such race conditions is to block all async. ETSs
+    %%     editing during the validation process, but this can result in big delays
+    %%     during the MIM initialisation phase.
+    %%   * also it's not clear how to interpret registration of the "prefix" based
+    %%     subdomain patterns, should we block the registration of the whole pattern
+    %%     or just only conflicting subdomains registration. blocking of the whloe
+    %%     pattern requires generation and verification of all the subdomains (with
+    %%     ETS blocking during that process), which depends on domains ETS size and
+    %%     might take too long.
+    %%   * and the last big issue with simple registration rejection approach, different
+    %%     nodes in the cluster might have different registration sequence, so we may
+    %%     end up in a situation when some nodes registered domain name as a subdomain,
+    %%     while other nodes registered it as a top level domain.
+    %%
+    %% the better way is to prohibit registration of top level domain if they are equal
+    %% to any of the FQDN subdomains or if beginning of domain name matches the prefix
+    %% of any subdomain template. in this case we don't need to verify subdomains at all,
+    %% verification of domain names against some limit number of subdomains patterns is
+    %% enough. and the only problem that we need to solve - mongooseim_domain_core must
+    %% be aware of all the subdomain patterns before it registers the first dynamic
+    %% domain. this would require minor configuration rework, e.g. tracking of subdomain
+    %% templates preprocessing (mongoose_subdomain_utils:make_subdomain_pattern/1 calls)
+    %% during toml config parsing.
+    Pattern1 = mongoose_subdomain_utils:make_subdomain_pattern("subdomain.@HOST@"),
+    Pattern2 = mongoose_subdomain_utils:make_subdomain_pattern("some.fqdn"),
+    Pattern3 = mongoose_subdomain_utils:make_subdomain_pattern("another.fqdn"),
+    Handler = mongoose_packet_handler:new(?MODULE),
+    %%------------------------------------------
+    %% testing type #1 subdomain names collision
```

Comment:
Here a well, one test case per collision type would be good. I would even split the check for type 1 in two.